### DISCUSSION TITLE
How to setup different exception_handler's for different routes or http_handlers?

### DISCUSSION THREAD

**Discussion by oObelix** [fastapi/fastapi#8396] [Questions] [answered]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from fastapi import FastAPI

app = FastAPI()


@app.exception_handler(RequestValidationError)
async def validation_exception(request, exc):
    return Response(
        content='Incorrect symbol',
        media_type="text/html",
        status_code=404
    )


@app.get('/')
def foo(q: str = Query(regex='^[0-9]+$')):
   pass

@app.post('/')
def bar(q: str = Query(regex='^[0-9]+$')):
    pass
```

### Description

I can add `@app.exception_handler(RequestValidationError)` for all Validation Errors like in code above. How to add different exception_handler's to get (foo()) and post (bar()) handlers with same regex validation?

```
from fastapi import FastAPI

app = FastAPI()

@app.get('/')
def foo(q: str = Query(regex='^[0-9]+$')):
   pass

@app.post('/')
def bar(q: str = Query(regex='^[0-9]+$')):
    pass

```

### Operating System

Linux

### FastAPI Version

0.85.1

### Python Version

3.10

**Comment by TheDavidFactor** [comment]
I have the exact same issue. I'm building a REST api, but one endpoint needs to be text/html, all others are application/json. I was able to change the request and response, but if there's a validation error it returns application/json. I don't want to make a global change because all other endpoints are application/json.

**Comment by JarroVGIT** [comment]
Well, out of the box you can't have different exception handlers for the same `Exception` subtype. That is because the exception handlers are registered at the upmost app level. 

One way of dealing with your requirement is to create a dependency function that takes the parameter as its parameter, but without the regex. This logic would then be moved to the function itself, and based on the request method you could raise a specific `Exception`. But that becomes unwieldy if your real-life code is a bit more complex than the example you gave (which I assume it is). 

On the other hand (and this is not helping you at all, I realise) you could argue that an error response from an API framework in the form of html is.. sketchy and should be JSON in the first place?

**Comment by yinziyan1206** [comment, marked as answer]
use different app, and mount it on main app. Its a basic usage in starlette

**Comment by TheDavidFactor** [comment]
@yinziyan1206 Thanks for the pointer. That's what I needed!

**Comment by TheDavidFactor** [comment]
Just came back to confirm this resolved my use case in case anyone stumbles across this in the future.
